File: CS50x/pset6/readability/readability.py
from cs50 import get_string


def main():
    # loop until length of string > 0
    while True:
        string = get_string("Text: ")
        if len(string) > 0:
            break

    index = calculate_colemanliau(string)
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


# count letters function
def countletters(string):
    nletters = 0
    for letter in string:
        if letter.isalpha():
            nletters += 1
    return nletters


# count words function
def countwords(string):
    nwords = 0
    for char in string:
        if char is " ":
            nwords += 1
    return nwords + 1


# count sentences function
def countsentences(string):
    nsentences = 0
    for char in string:
        if char in [".", "!", "?"]:
            nsentences += 1
    return nsentences


# and finally function to calculate the index from the Coleman-Liau formula:
# 0.0588 * L - 0.296 * S - 15.8
# L = average number of letters per 100 words
# S average number of sentences per 100 words
def calculate_colemanliau(string):
    L = (100 / countwords(string)) * countletters(string)
    S = (100 / countwords(string)) * countsentences(string)
    return round(0.0588 * L - 0.296 * S - 15.8)


main()

File: CS50x/pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int get_letters();
int get_words();
int get_sentences();
void calculation();

int main(void)
{
    // prompt user for a string/text:
    string text = get_string("Text: ");

    // get length of string
    int text_length = get_letters(text);

    // get number of words
    int text_words = get_words(text);

    // get number of sentences
    int text_sentences = get_sentences(text);

    // calculate the index
    calculation(text_words, text_length, text_sentences);
}

int get_letters(string input)
{
    // letter is any lowercase or uppercase character
    int text_length = strlen(input);
    int number_letters = 0;
    for (int i = 0; i < text_length; i++)
    {
        if (isalnum(input[i]))
        {
            number_letters++;
        }
    }
    return number_letters;
}

int get_words(string input)
{
    // letter is any lowercase or uppercase character
    int text_length = strlen(input);
    int n_words = 0;
    for (int i = 0; i < text_length; i++)
    {
        if (input[i] == ' ')
        {
            n_words++;
        }
    }
    return n_words + 1;
}

int get_sentences(string input)
{
    // sentences is any strings seperated by: ! . ?
    int text_length = strlen(input);
    int n_sentences = 0;
    for (int i = 0; i < text_length; i++)
    {
        if (input[i] == '!' || input[i] == '.' || input[i] == '?')
        {
            n_sentences++;
        }
    }
    return n_sentences;
}

void calculation(int words, int letters, int sentences)
{
    // calculates via the formula: index = 0.0588 * L - 0.296 * S - 15.8
    // L = average number of letters per 100 words
    // S = average number of sentences per 100 words
    double L = (100.0 / words) * letters;
    double S = (100.0 / words) * sentences;

    double index = 0.0588 * L - 0.296 * S - 15.8;
    int grade = round(index);

    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }
}